### ISSUE TITLE
Embedded where condition returned only subset of results

### ISSUE THREAD

**Issue by HuifangYeo** [atoti/atoti#464]
## Steps to reproduce
I have embedded `where` conditions as shown below:
```
m["interpolated_df"] = tt.where(
    l["Bucket"] == "Balance",
    1.0,                                                                          # true value
    tt.where(                                                                 # false embedded where condition
        l["PaymentDate"] == l["FlowDate"],                   
        tt.agg.mean(m["DF"]),                                         # true value
        tt.where(                                                              # false embedded where condition
            (l["PaymentDate"] > m["min_flow_date"]) & (l["PaymentDate"] < m["max_flow_date"]),
            m["_ccZeroRate"],                                           # true value
            tt.math.exp(-1 * m["rate_used"] * m["date_diff"] / 365),  # false value (theoretical defaulted value)
        )
    ),
)
```
Since I have multiple buckets and I have a defaulted value for the `false_value`, all my buckets should still be visible. However, with the first embedded condition, only the balance bucket remains.

![embedded_Where](https://user-images.githubusercontent.com/35210495/152337668-a34dcf39-c8a3-428c-9d4a-908f61682654.gif)

Source:  [fx_cashposition-missing-bucket.zip](https://github.com/atoti/atoti/files/7994130/fx_cashposition-missing-bucket.zip) 
(Refer to cell 32 and visualization after)

## Environment

- atoti: 0.6.4
- Python: 3.8.12
- Operating system: win32

**Comment by bloussou** [comment]
There is no column `asOfDate` in the `flows.csv`, is it the right values in the zip ?

**Comment by HuifangYeo** [comment]
sorry I missed this ticket.

I have a smaller example here: [measures_examples.zip](https://github.com/atoti/atoti/files/8057653/measures_examples.zip)
I have full running dates:
![image](https://user-images.githubusercontent.com/35210495/153795169-8f8f4230-6457-4659-bef2-9ecb143af62d.png)

However, once I have the embedded condition, only 1 row is returned.
![image](https://user-images.githubusercontent.com/35210495/153795231-bdb4b7e3-a84b-4d6f-8f43-f8b7e1da4342.png)

I should have value for every 1,15 of each months.

**Comment by bloussou** [comment]
Hello, I can't reproduce on the latest version of atoti on my computer. If possible can you try it ?

**Comment by fabiencelier** [comment]
Maybe fixed by the recent changes in the `where` ?
https://github.com/activeviam/atoti/pull/3165
@Acriss what do you think ?

**Comment by Acriss** [comment]
There is a chance, there has been a couple of internal changes to the `where` feature, first between 0.6.3 and 0.6.4 (where I may have introduced some issues), and now with 0.6.5-to-come.

I'll try to reproduce as well using 0.6.4 and latest, to confirm @bloussou 's comment.

**Comment by bloussou** [comment]
I have tested recreating a new notebook and copy/paste each cell to see if it happens on the continuous build and it is not. So I'm sure it will not be a bug in `0.6.5`.
When I copy/paste all the cells from the first one in a new notebook in `0.6.4` I didn't get the issue too. Can you test this too ?

**Comment by HuifangYeo** [comment]
Hi @bloussou / @Acriss,

I found the root of the issue. It's a matter of UI / MDX use problem:

__Scenario 1__: `Date parts` hierarchy not expanded.
Since the `where` conditions involve `Quarter` and `Day`, when the `Date parts` hierarchy is not expanded, nothing is returned.  
Users have no means of expanding the hierarchy either in this case, knowing only no data is returned when we know that there should be values if we expand the hierarchy.  
![scenario1](https://user-images.githubusercontent.com/35210495/153983932-97beef44-f6e7-4a1e-9a76-64f951105b4e.gif)

__Scenario 2__:  Added and expanded 1 month of the date hierarchy before adding `Seasonal increase`.
This is the reported case above. Since only 1 month is expanded, it's also the only one with members applicable for the `where` conditions and hence, the only one with results.
![scenario2](https://user-images.githubusercontent.com/35210495/153984429-40e7e5c3-38a6-4f96-a70a-a466b66dcd73.gif)

__Scenario 3__: Added the entire date hierarchy at the `Day` level. Lose the expandable behavior.
This is @bloussou use case when the date hierarchy is expanded and all members are available for the `where` condition.
![scenario3](https://user-images.githubusercontent.com/35210495/153984782-d6f6387d-5add-4d93-90e5-12c59325f87c.gif)

The key reason why I encountered this issue was that I wanted to have the expandable behaviour, in order to jump into the target month. However, it seems that from a UX point of view, I can't do this unless I have another measure that works on the `Year` level or I perform additional aggregation on the measure with the `where` condition.

**Comment by HuifangYeo** [comment]
Then again, this problem seems a bit different from the initial reported issue.

**Comment by HuifangYeo** [comment]
Yes, I agree it's more of a UX issue that can result in users missing values unknowingly if they do not understand how MDX functions in accordance with the expandable rows. It also took me a while to figure out when I have working solution from you.

**Comment by Acriss** [comment]
I think the issue stems from the fact that when you're expanding, in the pivot table, 2019 -> Quarter 1 -> Month 1 (without having selected the measure yet), we are implicitly creating some sort of filter that doesn't show anywhere.

Once you add the `where` measure, the set of queried location will be:

- [Date].[Date].[2018]
- [Date].[Date].[2019].[Quarter 1].[Month 1].[*]
- [Date].[Date].[2020]
- [Date].[Date].[2021]

However, the `where` measure only returns a result if the queried location expresses the level that's used in the condition (here, level `Days` is the one that is missing).
Thus, the MDX Query is actually ran on the given set of locations, but only returns for [Date].[Date].[2019].[Quarter 1].[Month 1].[*].

This is either a UX issue (showing the MDX editor, for instance, would have cleared any doubt, but it's hidden for Atoti if I'm not mistaken) or a misunderstanding of MDX, as you stated.

Note that in SQL, a query like `SELECT value.SUM WHERE Quarter == 1 AND DAY == 1 OR 15` wouldn't show any result either.
So this might as well come from the design choice of the `where` measure, that is : 
`where(l[Days] == 1, true_measure, false_measure)` does NOT return any result for [Date].[Date].[2018], because Days is not expressed.

This choice comes from the fact that we might not know at all how to aggregate `false_measure` at the Year level, and the results could have no sense at all. Thus, it's better not to give results. We expect that if the condition is on Days, the user doesn't expect anything at the Year level.. 

Even if we had an answer to this specific question, there is another one that we can't answer.
In the year 2018, there are Days that match the condition Quarter == 1 AND DAY == 1 OR 15 . For these days, the `where` evaluates to the `true_measure`. For the other days, it evaluates to the `false_measure`. How do we aggregate `false_measure` and `true_measure` together to create an aggregate for  [Date].[Date].[2018] ? 
Since we can't answer to this question, the behavior of the `where` measure cannot be tweaked.

Hope the explanation is satisfying :/

**Comment by HuifangYeo** [comment]
> Note that in SQL, a query like SELECT value.SUM WHERE Quarter == 1 AND DAY == 1 OR 15 wouldn't show any result either.
Try running the below two queries in [w3schools](https://www.w3schools.com/sql/trysql.asp?filename=trysql_select_all):
`SELECT * FROM Customers where country="Mexico" and postalcode in ("05021", "05023")`
or 
`SELECT * FROM Customers where country="Mexico" and (postalcode = "05021" or postalcode = "05023")`

Both of them do return results (Though they are not apple to apple comparison).

> For the other days, it evaluates to the false_measure. How do we aggregate false_measure and true_measure together to create an aggregate for [Date].[Date].[2018] ?

It was evaluated when we select the lowest level (`Days` level always present).
I'm not clear on the concept and maybe oversimplifying the topic. Just thinking if it's possible that when we have a `where` condition, for us to automatically expand on the level, in this case `Quarter` and `Days`?

**Comment by Acriss** [comment]
Yeah I shouldn't have added that SQL thingy. My bad.

> It was evaluated when we select the lowest level (`Days` level always present).

What I meant is, if your measure is `where(l["Day"] == 1, value.SUM, value.MEAN)`, this measure only makes sense when the Day level is expressed in the location. For the location `[Date].[Date].[2018]`, I cannot know what to do with the SUM on all days not  or 15, and the MEAN on days 1 and 15. So I wouldn't know what result to give for the specific location `[Date].[Date].[2018]`.
Thus, the location disappears.

> I'm not clear on the concept and maybe oversimplifying the topic. Just thinking if it's possible that when we have a `where` condition, for us to automatically expand on the level, in this case `Quarter` and `Days`?

But we cannot automatically select to expand 2018, 2020 and 2021 up to the days, simply because that is what was being done for 2019. ActiveUI generates the query without knowing it's a `where`, and simply asks for the measure you added in the columns, for the locations you expressed in the rows.
